daemon/nvidia_linux.go

```diff
@@ -30,19 +33,28 @@ var allNvidiaCaps = map[nvidia.Capability]struct{}{
 }
 
 func init() {
-	if _, err := exec.LookPath(nvidiaHook); err != nil {
-		// do not register Nvidia driver if helper binary is not present.
-		return
-	}
-	capset := capabilities.Set{"gpu": struct{}{}, "nvidia": struct{}{}}
-	nvidiaDriver := &deviceDriver{
-		capset:     capset,
-		updateSpec: setNvidiaGPUs,
-	}
-	for c := range allNvidiaCaps {
-		nvidiaDriver.capset[string(c)] = struct{}{}
+	// Register Nvidia driver if Nvidia helper binary is present.
+	// Else, register AMD driver if AMD helper binary is present.
+	if _, err := exec.LookPath(nvidiaHook); err == nil {
+		capset := capabilities.Set{"gpu": struct{}{}, "nvidia": struct{}{}}
+		nvidiaDriver := &deviceDriver{
+			capset:     capset,
+			updateSpec: setNvidiaGPUs,
+		}
+		for c := range allNvidiaCaps {
+			nvidiaDriver.capset[string(c)] = struct{}{}
+		}
+		registerDeviceDriver("nvidia", nvidiaDriver)
+	} else if _, err := exec.LookPath(amdContainerRuntimeExecutableName); err == nil {
+		capset := capabilities.Set{"gpu": struct{}{}, "amd": struct{}{}}
+		amdDriver := &deviceDriver{
+			capset:     capset,
+			updateSpec: setAMDGPUs,
+		}
+		registerDeviceDriver("amd", amdDriver)
+	} else {
+		// no "gpu" capability
 	}
```

Comment:
Perhaps an early return would work;

```suggestion
	if _, err := exec.LookPath(nvidiaHook); err == nil {
		capset := capabilities.Set{"gpu": struct{}{}, "nvidia": struct{}{}}
		for c := range allNvidiaCaps {
			capset[string(c)] = struct{}{}
		}
		registerDeviceDriver("nvidia",  &deviceDriver{
			capset:     capset,
			updateSpec: setNvidiaGPUs,
		})
		return
	}
	if _, err := exec.LookPath(amdContainerRuntimeExecutableName); err == nil {
		registerDeviceDriver("amd", &deviceDriver{
			capset:     capabilities.Set{"gpu": struct{}{}, "amd": struct{}{}},
			updateSpec: setAMDGPUs,
		})
		return
	}

	// no "gpu" capability
```

Reply:
Curious though; should `amd` and `nvidia` be considered mutually exclusive? Would splitting this into two `init` funcs (once in the nvidia file, one in the amd file) and both register a driver (if present) work?

Reply:
> Perhaps an early return would work;

@thaJeztah Changed as suggested.